File: src/main/scala/experiments/typeclasses/expressionProblem/Expr.scala
package experiments.typeclasses.expressionProblem

object Expr {

  trait Expr[E] {
    def eval(e: E): Int
  }
  implicit def addEvaluate[E: Expr](expr: E): Object {def evaluate: Int} = {
    new {
      def evaluate: Int = implicitly[Expr[E]].eval(expr)
    }
  }

  case class Number(value: Int)
  case class Plus[A: Expr, B: Expr](left: A, right: B)

  implicit def evalNumber = new Expr[Number] {
    def eval(e: Number) = e.value
  }
  implicit def evalPlus[A: Expr, B: Expr] = new Expr[Plus[A, B]] {
    def eval(e: Plus[A, B]) = implicitly[Expr[A]].eval(e.left) + implicitly[Expr[B]].eval(e.right)
  }
}

File: src/main/haskell/typeclasses/Expr.hs
{-# LANGUAGE ExistentialQuantification #-}
module Expr where

class Expr e where
    eval :: e -> Int

data Number = Number Int
data Plus a b = (Expr a, Expr b) => Plus a b

instance Expr Number where
    eval (Number n) = n

instance (Expr a, Expr b) => Expr (Plus a b) where
    eval (Plus x y) = (eval x) + (eval y)
